lib/mastodon/upgrade_cli.rb

```diff
@@ -41,23 +41,33 @@ def storage_schema
         klass.find_each do |record|
           attachment_names.each do |attachment_name|
             attachment = record.public_send(attachment_name)
+            success    = true
 
             next if attachment.blank? || attachment.storage_schema_version >= CURRENT_STORAGE_SCHEMA_VERSION
 
-            attachment.styles.each_key do |style|
-              case Paperclip::Attachment.default_options[:storage]
-              when :s3
-                upgrade_storage_s3(progress, attachment, style)
-              when :fog
-                upgrade_storage_fog(progress, attachment, style)
-              when :filesystem
-                upgrade_storage_filesystem(progress, attachment, style)
+            styles = attachment.styles.keys
+
+            styles << :original unless styles.include?(:original)
+
+            styles.each do |style|
+              success = success && begin
+                case Paperclip::Attachment.default_options[:storage]
+                when :s3
+                  upgrade_storage_s3(progress, attachment, style)
+                when :fog
+                  upgrade_storage_fog(progress, attachment, style)
+                when :filesystem
+                  upgrade_storage_filesystem(progress, attachment, style)
+                end
               end
 
               progress.increment
             end
 
-            attachment.instance_write(:storage_schema_version, CURRENT_STORAGE_SCHEMA_VERSION)
+            # Hopefully if this does happen, it happens for all styles, but if
+            # transfer of one or more of the styles fails, we should not mark
+            # the record as upgraded
+            attachment.instance_write(:storage_schema_version, CURRENT_STORAGE_SCHEMA_VERSION) if success
```

Comment:
If an error occurs for one style and not the others, we'll have some files already moved, and re-running the task will always fail, right? Is there something sensible we can do?

Reply:
Hm, what if only `:original` will determine whether upgrade failed or not, and the other styles will be ignored? Because you could re-generate other styles from `:original` as long as you had it.

Reply:
Hm, ok, but we don't have anything to actually re-generate other styles, do we?

Reply:
That's what paperclip's `reprocess!` does

Reply:
ok, but there's nothing for an admin to run it, and it wouldn't get triggered automatically either, would it?

Reply:
No, but I still consider it fixable, e.g. you'd give the admin code to run in console.